File: src/services/FederatedLearningService.js
import * as FileSystem from 'expo-file-system';
import SecurityService from './SecurityService';
import aiConfig from '../../config/ai-config.json';

class FederatedLearningService {
  constructor() {
    this.feedbackQueue = [];
    this.lastUpdate = null;
    this.isEnabled = aiConfig.federated_learning.enabled;
  }

  async initialize() {
    if (!this.isEnabled) {
      console.log('Federated learning devre dışı');
      return;
    }

    try {
      // Geri bildirim kuyruğunu yükle
      await this.loadFeedbackQueue();

      console.log('Federated learning servisi başlatıldı');
    } catch (error) {
      console.error('Federated learning başlatma hatası:', error);
    }
  }

  async addFeedback(analysisResult, userFeedback) {
    if (!this.isEnabled) {
      return;
    }

    try {
      const feedback = {
        id: Date.now().toString(),
        timestamp: Date.now(),
        analysisResult: analysisResult,
        userFeedback: userFeedback,
        deviceInfo: await this.getAnonymousDeviceInfo()
      };

      // Geri bildirimi şifrele
      const encryptedFeedback = await SecurityService.encryptData(feedback);

      this.feedbackQueue.push(encryptedFeedback);

      // Kuyruk boyutu kontrolü
      if (this.feedbackQueue.length > 50) {
        this.feedbackQueue = this.feedbackQueue.slice(-50); // Son 50'yi tut
      }

      await this.saveFeedbackQueue();

      console.log('Geri bildirim eklendi:', feedback.id);
    } catch (error) {
      console.error('Geri bildirim ekleme hatası:', error);
    }
  }

  async getAnonymousDeviceInfo() {
    // Anonim cihaz bilgileri
    return {
      platform: 'mobile', // Genel platform bilgisi
      appVersion: '1.0.0',
      modelVersion: aiConfig.models.baby_sound_classifier.version,
      region: 'TR', // Anonim bölge bilgisi
      // Kişisel tanımlayıcı bilgiler içermez
    };
  }

  async processLocalModelUpdate() {
    if (!this.isEnabled || this.feedbackQueue.length === 0) {
      return;
    }

    try {
      // Yerel model güncellemesi
      // Bu kısımda geri bildirimler kullanılarak model iyileştirilecek

      const recentFeedback = this.feedbackQueue.slice(-10); // Son 10 geri bildirim

      // Model parametrelerini güncelle
      await this.updateModelParameters(recentFeedback);

      // Güncelleme zamanını kaydet
      this.lastUpdate = Date.now();
      await this.saveUpdateTimestamp();

      console.log('Yerel model güncellendi');
    } catch (error) {
      console.error('Yerel model güncelleme hatası:', error);
    }
  }

  async updateModelParameters(feedbackBatch) {
    // Model parametrelerini geri bildirimlere göre güncelle
    // Bu kısımda federated learning algoritması çalışacak

    for (const encryptedFeedback of feedbackBatch) {
      try {
        const feedback = await SecurityService.decryptData(encryptedFeedback);

        // Model parametrelerini ayarla
        // TensorFlow.js ile model güncelleme

        console.log('Geri bildirim işlendi:', feedback.id);
      } catch (error) {
        console.error('Geri bildirim işleme hatası:', error);
      }
    }
  }

  async prepareAnonymousUpdate() {
    if (!this.isEnabled || !this.lastUpdate) {
      return null;
    }

    try {
      // Son güncellemeden beri ne kadar zaman geçtiğini kontrol et
      const daysSinceUpdate = (Date.now() - this.lastUpdate) / (24 * 60 * 60 * 1000);

      if (daysSinceUpdate < aiConfig.federated_learning.update_interval_days) {
        return null;
      }

      // Anonim model güncellemesi hazırla
      const updatePackage = {
        id: Date.now().toString(),
        timestamp: Date.now(),
        deviceInfo: await this.getAnonymousDeviceInfo(),
        modelVersion: aiConfig.models.baby_sound_classifier.version,
        // Anonimleştirilmiş model güncellemeleri
        parameters: await this.getAnonymousModelParameters()
      };

      const encryptedUpdate = await SecurityService.encryptData(updatePackage);

      return encryptedUpdate;
    } catch (error) {
      console.error('Anonim güncelleme hazırlama hatası:', error);
      return null;
    }
  }

  async getAnonymousModelParameters() {
    // Anonimleştirilmiş model parametreleri
    // Gerçek model parametrelerini anonim hale getir
    return {
      parameterCount: 1000, // Demo değeri
      updateSize: '2.3KB', // Demo değeri
      accuracy: 0.85, // Demo değeri
      // Gerçek parametreler anonimleştirilmiş formatta
    };
  }

  async saveFeedbackQueue() {
    try {
      const queuePath = FileSystem.documentDirectory + 'federated/feedback_queue.json';

      // Dizini oluştur
      const dirInfo = await FileSystem.getInfoAsync(FileSystem.documentDirectory + 'federated/');
      if (!dirInfo.exists) {
        await FileSystem.makeDirectoryAsync(FileSystem.documentDirectory + 'federated/');
      }

      await FileSystem.writeAsStringAsync(
        queuePath,
        JSON.stringify(this.feedbackQueue),
        { encoding: FileSystem.EncodingType.UTF8 }
      );
    } catch (error) {
      console.error('Geri bildirim kuyruğu kaydetme hatası:', error);
    }
  }

  async loadFeedbackQueue() {
    try {
      const queuePath = FileSystem.documentDirectory + 'federated/feedback_queue.json';
      const fileInfo = await FileSystem.getInfoAsync(queuePath);

      if (fileInfo.exists) {
        const queueData = await FileSystem.readAsStringAsync(queuePath);
        this.feedbackQueue = JSON.parse(queueData);
      }
    } catch (error) {
      console.error('Geri bildirim kuyruğu yükleme hatası:', error);
      this.feedbackQueue = [];
    }
  }

  async saveUpdateTimestamp() {
    try {
      const timestampPath = FileSystem.documentDirectory + 'federated/last_update.json';
      const dirInfo = await FileSystem.getInfoAsync(FileSystem.documentDirectory + 'federated/');

      if (!dirInfo.exists) {
        await FileSystem.makeDirectoryAsync(FileSystem.documentDirectory + 'federated/');
      }

      await FileSystem.writeAsStringAsync(
        timestampPath,
        JSON.stringify({ lastUpdate: this.lastUpdate }),
        { encoding: FileSystem.EncodingType.UTF8 }
      );
    } catch (error) {
      console.error('Güncelleme zamanı kaydetme hatası:', error);
    }
  }

  async loadUpdateTimestamp() {
    try {
      const timestampPath = FileSystem.documentDirectory + 'federated/last_update.json';
      const fileInfo = await FileSystem.getInfoAsync(timestampPath);

      if (fileInfo.exists) {
        const timestampData = await FileSystem.readAsStringAsync(timestampPath);
        const parsed = JSON.parse(timestampData);
        this.lastUpdate = parsed.lastUpdate;
      }
    } catch (error) {
      console.error('Güncelleme zamanı yükleme hatası:', error);
    }
  }

  async cleanup() {
    this.feedbackQueue = [];
    this.lastUpdate = null;
  }

  getServiceInfo() {
    return {
      enabled: this.isEnabled,
      feedbackCount: this.feedbackQueue.length,
      lastUpdate: this.lastUpdate,
      anonymousOnly: aiConfig.federated_learning.anonymous_only,
      updateInterval: aiConfig.federated_learning.update_interval_days
    };
  }
}

export default new FederatedLearningService();

File: lib/services/federated_learning_service.dart
import 'dart:convert';
import 'dart:typed_data';
import 'dart:math';
import 'package:crypto/crypto.dart';
import 'package:uuid/uuid.dart';

import '../models/user_feedback.dart';
import '../models/cry_analysis.dart';
import 'storage_service.dart';
import 'ai_service.dart';

/// Federated Learning servisi - Kullanıcı verilerini cihazda tutarak
/// model iyileştirmesi sağlar
class FederatedLearningService {
  static const String _deviceId = 'device_id';
  static const int _minFeedbackForUpdate = 10;
  static const int _maxFeedbackBatch = 50;
  static const Duration _updateInterval = Duration(days: 7);
  
  static bool _isInitialized = false;
  static String? _uniqueDeviceId;
  static DateTime? _lastUpdate;
  
  /// Servisi başlat
  static Future<void> initialize() async {
    if (_isInitialized) return;
    
    try {
      // Benzersiz cihaz ID'si oluştur veya yükle
      _uniqueDeviceId = StorageService.getSetting<String>(_deviceId);
      if (_uniqueDeviceId == null) {
        _uniqueDeviceId = const Uuid().v4();
        await StorageService.saveSetting(_deviceId, _uniqueDeviceId);
      }
      
      // Son güncelleme zamanını yükle
      final lastUpdateStr = StorageService.getSetting<String>('last_fl_update');
      if (lastUpdateStr != null) {
        _lastUpdate = DateTime.parse(lastUpdateStr);
      }
      
      _isInitialized = true;
      print('Federated Learning Service başlatıldı');
      
      // Otomatik güncelleme kontrolü
      _schedulePeriodicUpdates();
    } catch (e) {
      print('Federated Learning başlatma hatası: $e');
      rethrow;
    }
  }
  
  /// Kullanıcı geri bildirimiyle modeli yerel olarak güncelle
  static Future<void> updateModelWithFeedback(FeedbackData feedback) async {
    if (!_isInitialized) await initialize();
    
    try {
      // Geri bildirim verilerini hazırla
      final trainingData = await _prepareFeedbackForTraining([feedback]);
      
      // Yerel model güncellemesi
      await _performLocalModelUpdate(trainingData);
      
      // Geri bildirimi eğitimde kullanıldı olarak işaretle
      await StorageService.markFeedbackAsUsed(feedback.id);
      
      print('Model geri bildirimle güncellendi: ${feedback.id}');
    } catch (e) {
      print('Model güncelleme hatası: $e');
    }
  }
  
  /// Toplu geri bildirimlerle model güncelleme
  static Future<void> updateModelWithBatchFeedback() async {
    if (!_isInitialized) await initialize();
    
    try {
      // Eğitimde kullanılmamış geri bildirimleri al
      final pendingFeedback = StorageService.getFeedbackForTraining();
      
      if (pendingFeedback.length < _minFeedbackForUpdate) {
        print('Yetersiz geri bildirim sayısı: ${pendingFeedback.length}');
        return;
      }
      
      // Batch boyutunu sınırla
      final feedbackBatch = pendingFeedback.take(_maxFeedbackBatch).toList();
      
      // Eğitim verilerini hazırla
      final trainingData = await _prepareFeedbackForTraining(feedbackBatch);
      
      // Yerel model güncellemesi
      await _performLocalModelUpdate(trainingData);
      
      // Kullanılan geri bildirimleri işaretle
      for (final feedback in feedbackBatch) {
        await StorageService.markFeedbackAsUsed(feedback.id);
      }
      
      // Son güncelleme zamanını kaydet
      _lastUpdate = DateTime.now();
      await StorageService.saveSetting('last_fl_update', _lastUpdate!.toIso8601String());
      
      print('Model ${feedbackBatch.length} geri bildirimle güncellendi');
      
      // Anonimleşmiş model güncellemelerini hazırla (opsiyonel)
      await _prepareAnonymizedModelUpdate(trainingData);
    } catch (e) {
      print('Toplu model güncelleme hatası: $e');
    }
  }
  
  /// Geri bildirim verilerini eğitim formatına çevir
  static Future<TrainingData> _prepareFeedbackForTraining(List<FeedbackData> feedbackList) async {
    final correctPredictions = <TrainingExample>[];
    final incorrectPredictions = <TrainingExample>[];
    
    for (final feedback in feedbackList) {
      try {
        // İlgili analizi bul
        final analyses = StorageService.getAllAnalyses();
        final analysis = analyses.firstWhere((a) => a.id == feedback.analysisId);
        
        final example = TrainingExample(
          features: analysis.features,
          predictedType: analysis.cryType,
          actualType: feedback.correctedType ?? analysis.cryType,
          confidence: analysis.confidence,
          satisfactionRating: feedback.satisfactionRating,
          wasAccurate: feedback.wasAccurate,
        );
        
        if (feedback.wasAccurate) {
          correctPredictions.add(example);
        } else {
          incorrectPredictions.add(example);
        }
      } catch (e) {
        print('Geri bildirim hazırlama hatası: $e');
        continue;
      }
    }
    
    return TrainingData(
      correctPredictions: correctPredictions,
      incorrectPredictions: incorrectPredictions,
      deviceId: _uniqueDeviceId!,
      timestamp: DateTime.now(),
    );
  }
  
  /// Yerel model güncellemesi gerçekleştir
  static Future<void> _performLocalModelUpdate(TrainingData trainingData) async {
    try {
      // Eğitim istatistiklerini hesapla
      final stats = _calculateTrainingStatistics(trainingData);
      
      // Model ağırlık güncellemelerini hesapla
      final weightUpdates = _calculateWeightUpdates(trainingData);
      
      // AI servisine model güncelleme bilgilerini gönder
      await AIService.updateModelWithFeedback(
        trainingData.correctPredictions.map((e) => FeedbackData(
          id: const Uuid().v4(),
          analysisId: '',
          wasAccurate: true,
          correctedType: e.actualType,
          satisfactionRating: e.satisfactionRating,
          suggestionWasHelpful: true,
          createdAt: DateTime.now(),
        )).toList() + 
        trainingData.incorrectPredictions.map((e) => FeedbackData(
          id: const Uuid().v4(),
          analysisId: '',
          wasAccurate: false,
          correctedType: e.actualType,
          satisfactionRating: e.satisfactionRating,
          suggestionWasHelpful: false,
          createdAt: DateTime.now(),
        )).toList()
      );
      
      // Güncelleme istatistiklerini kaydet
      await _saveUpdateStatistics(stats);
      
      print('Yerel model güncellendi: ${stats.totalSamples} örnek');
    } catch (e) {
      print('Yerel model güncelleme hatası: $e');
      rethrow;
    }
  }
  
  /// Eğitim istatistiklerini hesapla
  static TrainingStatistics _calculateTrainingStatistics(TrainingData trainingData) {
    final totalSamples = trainingData.correctPredictions.length + trainingData.incorrectPredictions.length;
    final accuracy = trainingData.correctPredictions.length / totalSamples;
    
    // Ağlama türü dağılımı
    final typeDistribution = <CryType, int>{};
    for (final example in [...trainingData.correctPredictions, ...trainingData.incorrectPredictions]) {
      typeDistribution[example.actualType] = (typeDistribution[example.actualType] ?? 0) + 1;
    }
    
    // Ortalama memnuniyet skoru
    final satisfactionScores = [...trainingData.correctPredictions, ...trainingData.incorrectPredictions]
        .map((e) => e.satisfactionRating.toDouble()).toList();
    final averageSatisfaction = satisfactionScores.reduce((a, b) => a + b) / satisfactionScores.length;
    
    return TrainingStatistics(
      totalSamples: totalSamples,
      accuracy: accuracy,
      typeDistribution: typeDistribution,
      averageSatisfaction: averageSatisfaction,
      timestamp: DateTime.now(),
    );
  }
  
  /// Model ağırlık güncellemelerini hesapla
  static Map<String, List<double>> _calculateWeightUpdates(TrainingData trainingData) {
    // Basitleştirilmiş ağırlık güncelleme algoritması
    // Gerçek uygulamada daha karmaşık federated averaging kullanılır
    
    final weightUpdates = <String, List<double>>{};
    
    // Doğru tahminler için pozitif güncelleme
    for (final example in trainingData.correctPredictions) {
      final typeIndex = example.actualType.index;
      final updateKey = 'layer_output_$typeIndex';
      
      if (!weightUpdates.containsKey(updateKey)) {
        weightUpdates[updateKey] = List.filled(6, 0.0); // 6 ağlama türü
      }
      
      // Güven skoruna göre pozitif güncelleme
      weightUpdates[updateKey]![typeIndex] += example.confidence * 0.1;
    }
    
    // Yanlış tahminler için negatif güncelleme
    for (final example in trainingData.incorrectPredictions) {
      final predictedIndex = example.predictedType.index;
      final actualIndex = example.actualType.index;
      
      final updateKey = 'layer_output_$predictedIndex';
      if (!weightUpdates.containsKey(updateKey)) {
        weightUpdates[updateKey] = List.filled(6, 0.0);
      }
      
      // Yanlış tahmini azalt, doğru tahmini artır
      weightUpdates[updateKey]![predictedIndex] -= example.confidence * 0.05;
      weightUpdates[updateKey]![actualIndex] += example.confidence * 0.05;
    }
    
    return weightUpdates;
  }
  
  /// Anonimleşmiş model güncellemelerini hazırla
  static Future<void> _prepareAnonymizedModelUpdate(TrainingData trainingData) async {
    try {
      // Kişisel verileri çıkar, sadece model parametrelerini tut
      final anonymizedUpdate = {
        'device_hash': _hashDeviceId(_uniqueDeviceId!),
        'update_weights': _calculateWeightUpdates(trainingData),
        'sample_count': trainingData.correctPredictions.length + trainingData.incorrectPredictions.length,
        'accuracy': trainingData.correctPredictions.length / 
            (trainingData.correctPredictions.length + trainingData.incorrectPredictions.length),
        'timestamp': DateTime.now().toIso8601String(),
      };
      
      // Anonimleşmiş güncellemeyi yerel olarak sakla
      // Gerçek uygulamada bu güncelleme sunucuya gönderilebilir
      await StorageService.saveSetting('last_anonymized_update', jsonEncode(anonymizedUpdate));
      
      print('Anonimleşmiş model güncellemesi hazırlandı');
    } catch (e) {
      print('Anonimleşmiş güncelleme hatası: $e');
    }
  }
  
  /// Güncelleme istatistiklerini kaydet
  static Future<void> _saveUpdateStatistics(TrainingStatistics stats) async {
    final statsData = {
      'total_samples': stats.totalSamples,
      'accuracy': stats.accuracy,
      'type_distribution': stats.typeDistribution.map((k, v) => MapEntry(k.name, v)),
      'average_satisfaction': stats.averageSatisfaction,
      'timestamp': stats.timestamp.toIso8601String(),
    };
    
    await StorageService.saveSetting('fl_statistics', jsonEncode(statsData));
  }
  
  /// Cihaz ID'sini hash'le (gizlilik için)
  static String _hashDeviceId(String deviceId) {
    final bytes = utf8.encode(deviceId + 'baby_cry_salt');
    final digest = sha256.convert(bytes);
    return digest.toString().substring(0, 16); // İlk 16 karakter
  }
  
  /// Periyodik güncellemeleri planla
  static void _schedulePeriodicUpdates() {
    // Her gün kontrol et
    Stream.periodic(const Duration(hours: 24)).listen((_) async {
      if (_shouldPerformUpdate()) {
        await updateModelWithBatchFeedback();
      }
    });
  }
  
  /// Güncelleme yapılıp yapılmayacağını kontrol et
  static bool _shouldPerformUpdate() {
    if (_lastUpdate == null) return true;
    
    final timeSinceLastUpdate = DateTime.now().difference(_lastUpdate!);
    return timeSinceLastUpdate >= _updateInterval;
  }
  
  /// Federated Learning istatistiklerini getir
  static Map<String, dynamic> getStatistics() {
    try {
      final statsStr = StorageService.getSetting<String>('fl_statistics');
      if (statsStr != null) {
        return jsonDecode(statsStr);
      }
      return {};
    } catch (e) {
      print('İstatistik getirme hatası: $e');
      return {};
    }
  }
  
  /// Servisi temizle
  static void dispose() {
    _isInitialized = false;
    _uniqueDeviceId = null;
    _lastUpdate = null;
  }
}

/// Eğitim verisi sınıfı
class TrainingData {
  final List<TrainingExample> correctPredictions;
  final List<TrainingExample> incorrectPredictions;
  final String deviceId;
  final DateTime timestamp;
  
  TrainingData({
    required this.correctPredictions,
    required this.incorrectPredictions,
    required this.deviceId,
    required this.timestamp,
  });
}

/// Eğitim örneği sınıfı
class TrainingExample {
  final Map<String, double> features;
  final CryType predictedType;
  final CryType actualType;
  final double confidence;
  final int satisfactionRating;
  final bool wasAccurate;
  
  TrainingExample({
    required this.features,
    required this.predictedType,
    required this.actualType,
    required this.confidence,
    required this.satisfactionRating,
    required this.wasAccurate,
  });
}

/// Eğitim istatistikleri sınıfı
class TrainingStatistics {
  final int totalSamples;
  final double accuracy;
  final Map<CryType, int> typeDistribution;
  final double averageSatisfaction;
  final DateTime timestamp;
  
  TrainingStatistics({
    required this.totalSamples,
    required this.accuracy,
    required this.typeDistribution,
    required this.averageSatisfaction,
    required this.timestamp,
  });
}